### PR TITLE
chore: take-home exam submission

### PR DESCRIPTION
**Pull request by 87xie**
## 💡 Background and solution

I am using the group by feature approach for folder structure which helps me find related files easily.

Below is the component hierarchy overview.

```mermaid
flowchart TB
  editor-->editor-provider
  editor-provider-->image-group
  editor-provider-->code-block
  subgraph unuploaded
    image-dropzone
  end
  subgraph uploaded
    image-canvas
    image-selection
  end
  image-group-->unuploaded
  image-group-->uploaded
```

### Why use the context provider pattern

- It's not a complex state, Most of the re-render is necessary when the state change.
- Friendly for component composition. Keep the exposed context API consistency to help alternate the sub-component or better implementation easily.

If resize or reposition the selection block needs to renew the preview data immediately, Maybe can split the context or use the state management library for the performance issue.

### Canvas

- Most of the canvas view updates do not rely on react manipulation, so I use `useRef` to keep the canvas internal state, e.g. mouse current status.
- Saving the latest canvas DOMRect via mousedown and use in mousemove, avoid get box metrics frequently (might cause unnecessary reflow).

### Libaraies

- react-dropzone: For file upload, prop getters API looks good to me.
- react-rnd: It's a higher-level wrapper of react-draggable and re-resizable combination, I think the API interface is easy to use.
- react-icons: Styling.
- styled-components: Styling.
- polished: CSS-in-JS toolset. use for setup the normalize CSS.
- nanoid: Unique string id for selection block.

## 🤔 TODO

- Naming is hard. If there are better ones please let me know.
- Not familiar with typescript, would you provide the suggestion for code smells?
- Maybe some grammar mistakes in this pull request description, hope it doesn't affect your reading.

## ☑️ Self Check before Merge

- [ ] Doc and Wiki are updated/provided or not needed.
- [ ] Followed the Commit Message Guidelines listed in CONTRIBUTING.md

### CODE REVIEW

File: src/features/image-map-editor/image-dropzone.tsx [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    acceptedFiles,
+    getRootProps,
+    getInputProps,
+  } = useDropzone(dropzoneOptions);
+  const { dispatch } = useEditor();
+
+  const addImg = useCallback((file) => {
+    const img = new Image();
+    img.src = URL.createObjectURL(file);
+    img.src = URL.createObjectURL(file);
```
Comment (Joseph7451797) [inline src/features/image-map-editor/image-dropzone.tsx:25]: Is this line a duplicate one?

File: src/features/image-map-editor/image-dropzone.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+      dispatch({ type: 'add-image', imgData })
+    };
+  }, [dispatch]);
+
+  useEffect(() => {
+    if (!acceptedFiles.length) return;
+    const [file] = acceptedFiles;
+    if (!file) return;
+    addImg(file);
+  }, [acceptedFiles, addImg]);
```
Comment (Joseph7451797) [inline src/features/image-map-editor/image-dropzone.tsx:42]: In the react-dropzone's documents, they use `onDrop` as the useDropzone's parameter to handle files. I'd like to know why you handle files in the effect function.

File: src/features/image-map-editor/image-canvas.tsx [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+  onDrawStart?: DrawStartCallback;
+  onDrawStop?: DrawStopCallback;
+}
+
+const ImageCanvas = ({
+  width,
+  height,
+  onDrawStart,
+  onDrawStop,
+  ...rest
```
Comment (Joseph7451797) [inline src/features/image-map-editor/image-canvas.tsx:25]: I think this is an empty object. Any reason to pass it down?

File: src/features/image-map-editor/image-dropzone.tsx (hunk 1 above)
Comment (87xie) [reply src/features/image-map-editor/image-dropzone.tsx:25]: Ah yes, need to clean up. Thanks for reviewing.

**Comment by Joseph7451797** [comment]
One more question: I saw some functions are wrapped with `useCallback`, and some aren't. Can you explain why?

File: src/features/image-map-editor/image-dropzone.tsx (hunk 2 above)
Comment (87xie) [reply src/features/image-map-editor/image-dropzone.tsx:42]: I didn't notice the `onDropAccepted` callback option. I will refactor it.

File: src/features/image-map-editor/image-canvas.tsx (hunk 3 above)
Comment (87xie) [reply src/features/image-map-editor/image-canvas.tsx:25]: I set the data attribute on the element and styling on the parent scope.
Should I remove the `...rest` piece for a better understanding?

**Comment by 87xie** [comment]
> One more question: I saw some functions are wrapped with `useCallback`, and some aren't. Can you explain why?

I only used it when I need to put the callback as a dependency, keep simple if there are no re-render performance issues.

File: src/features/image-map-editor/image-canvas.tsx (hunk 3 above)
Comment (87xie) [reply src/features/image-map-editor/image-canvas.tsx:25]: Thanks for the suggestion, it's clear and helpful.

**Comment by Joseph7451797** [comment]
Reviewed.

**Pull request closed without merging**